### PR TITLE
late customization API

### PR DESCRIPTION
**Pull request by washort**

### CODE REVIEW

**Comment by chuckharmston** [comment]
Drive-by: can we put it all in a new app? This actually has nothing to do with feed.

File: mkt/feed/serializers.py [hunk 1]
```diff
@@ -485,3 +485,20 @@ def fake_object(self, data):
                     item_type)))
 
         return feed_item
+
+
+class LateCustomizationSerializer(URLSerializerMixin,
+                                  serializers.ModelSerializer):
+    """
+    A serializer for the FeedItem class, which wraps all items that live on the
```
Comment (diox) [inline mkt/feed/serializers.py]: confusing docstring, this doesn't seem feed-related

**Comment by diox** [comment]
+1 on @chuckharmston suggestion, I was confused as to why this is called feed initially.

**Comment by diox** [comment]
Also, bug number ?

File: mkt/feed/constants.py [hunk 2]
```diff
@@ -94,6 +94,7 @@
 FEED_TYPE_BRAND = 'brand'
 FEED_TYPE_COLL = 'collection'
 FEED_TYPE_SHELF = 'shelf'
+FEED_TYPE_LATE_CUSTOMIZATION = 'late-customization'
```
Comment (diox) [inline mkt/feed/constants.py]: unneeded ?

Comment (washort) [inline mkt/feed/constants.py]: oops, a leftover from my earlier bad idea :)

File: mkt/latecustomization/migrations/0001_initial.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+import mkt.constants.carriers
+import mkt.constants.regions
```
Comment (diox) [inline mkt/latecustomization/migrations/0001_initial.py]: Doing mkt imports in migrations is dangerous, as it can attempt to load models before they are ready and break the migration later when changes to that model are introduced in the future. I'm curious, did django generate this one or did you alter it by hand ?

File: mkt/latecustomization/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                ('carrier', models.IntegerField(choices=[(15, mkt.constants.carriers.TELENOR), (7, mkt.constants.carriers.KDDI), (11, mkt.constants.carriers.SINGTEL), (13, mkt.constants.carriers.SPRINT), (14, mkt.constants.carriers.TELECOM_ITALIA_GROUP), (17, mkt.constants.carriers.VIMPELCOM), (9, mkt.constants.carriers.MEGAFON), (2, mkt.constants.carriers.AMERICA_MOVIL), (6, mkt.constants.carriers.HUTCHINSON_THREE_GROUP), (19, mkt.constants.carriers.CONGSTAR), (22, mkt.constants.carriers.ORANGE), (21, mkt.constants.carriers.MTN), (0, mkt.constants.carriers.UNKNOWN_CARRIER), (20, mkt.constants.carriers.O2), (12, mkt.constants.carriers.SMART), (5, mkt.constants.carriers.ETISALAT), (18, mkt.constants.carriers.GRAMEENPHONE), (8, mkt.constants.carriers.KT), (10, mkt.constants.carriers.QTEL), (3, mkt.constants.carriers.CHINA_UNICOM), (1, mkt.constants.carriers.TELEFONICA), (16, mkt.constants.carriers.TMN), (4, mkt.constants.carriers.DEUTSCHE_TELEKOM)])),
```
Comment (diox) [inline mkt/latecustomization/migrations/0001_initial.py]: Related: I wish we had a solution for this, it's a bit ridiculous to have the choices in the migration when we know the database is not going to use them.

File: docs/api/topics/latecustomization.rst [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+.. http:post:: /api/v2/latecustomization/
+
+  Set the late-customization list for a given region and carrier. Replaces
+  existing list.
+
+  **Request**
+
+  :param app: A webapp ID.
+  :type app: int
+  :param region: A region ID.
```
Comment (diox) [inline docs/api/topics/latecustomization.rst]: The API (and tests) appear to be accepting slugs, not IDs.

File: docs/api/topics/latecustomization.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+.. http:get:: /api/v2/latecustomization/?carrier=(str:carrier)&region=(str:region)
+
+  Returns a list of late customization items for this carrier/region.
+
+  **Response**
+
+  :param app: A webapp ID.
+  :type app: int
+  :param region: A region ID.
+  :type region: int
```
Comment (diox) [inline docs/api/topics/latecustomization.rst]: Same as below, the API seems to be returning slugs, not IDs (a good thing IMHO, but docs should be updated)

File: mkt/latecustomization/views.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+from mkt.constants.carriers import CARRIER_MAP
+from mkt.constants.regions import REGIONS_DICT
+from mkt.feed.permissions import FeedPermission
+from mkt.feed.views import FeedShelfPermissionMixin, RegionCarrierFilter
+from mkt.latecustomization.models import LateCustomizationItem
+from mkt.latecustomization.serializers import LateCustomizationSerializer
+
+
+class LateCustomizationViewSet(FeedShelfPermissionMixin, CORSMixin,
+                               viewsets.ModelViewSet):
```
Comment (diox) [inline mkt/latecustomization/views.py]: `ModelViewSet` has all the HTTP verbs, but you don't seem to want them ? (They are not documented / tested). From the docs it looks like you only need `GenericViewSet` and `mixins.DestroyModelMixin` ?

Comment (diox) [inline mkt/latecustomization/views.py]: Also, you're missing `MarketplaceView`, so you don't get the exception handling and the tastypie-style pagination that the rest of the APIs have.

File: mkt/latecustomization/views.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    Operations for late customization items.
+    """
+    authentication_classes = [RestOAuthAuthentication,
+                              RestSharedSecretAuthentication]
+
+    permission_classes = [FeedPermission]
```
Comment (diox) [inline mkt/latecustomization/views.py]: Seems weird that you use `FeedShelfPermissionMixin` and `FeedPermission` outside of the feed when it doesn't seem related to feed.

File: mkt/latecustomization/views.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        return super(LateCustomizationViewSet, self).list(request, *args,
+                                                          **kwargs)
+
+    def create(self, request, *args, **kwargs):
+        data = self.req_data()
+        try:
+            self.require_operator_permission(
+                request.user, data['carrier'], data['region'])
+        except (KeyError, MultiValueDictKeyError):
+            raise ParseError()
```
Comment (diox) [inline mkt/latecustomization/views.py]: Return a meaningful error message ?

File: mkt/latecustomization/migrations/0001_initial.py (hunk 4 above)
Comment (washort) [inline mkt/latecustomization/migrations/0001_initial.py]: We could make a custom field that doesn't serialize this info in the migrations.

File: mkt/latecustomization/migrations/0001_initial.py (hunk 3 above)
Comment (washort) [inline mkt/latecustomization/migrations/0001_initial.py]: This was generated.

Comment (diox) [inline mkt/latecustomization/migrations/0001_initial.py]: Damn, that's bad. Those imports should be ok, but that was close.

File: mkt/latecustomization/views.py (hunk 8 above)
Comment (washort) [inline mkt/latecustomization/views.py]: since this is for use by carriers it seems to be the right permission checks.

Comment (chuckharmston) [inline mkt/latecustomization/views.py]: It should use https://github.com/mozilla/zamboni/blob/master/mkt/operators/permissions.py#L6

Comment (diox) [inline mkt/latecustomization/views.py]: yeah `IsOperatorPermission` seems more suited to that

File: mkt/urls.py [hunk 10]
```diff
@@ -41,6 +42,12 @@
         mini_langpack_manifest, name='langpack.manifest'),
     url('^extension/(?P<uuid>[0-9a-f]{32})/manifest.json$',
         mini_extension_manifest, name='extension.mini_manifest'),
+    url('^late_customization/%s/manifest.webapp$' % mkt.ADDON_UUID,
```
Comment (diox) [inline mkt/urls.py]: So, we have a different mini-manifest. Is platform doing some transformation ? Or are we going to add a hack to fireplace to have it consider the app is installed no matter what manifest is used ?

**Comment by washort** [comment]
Now with support for addons. re-r?

File: docs/api/topics/latecustomization.rst [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  :status 200: Successfully completed.
+
+
+  .. code-block:: json
+
+      {
+          "objects": [
+              {
+                  "app": "carrier-provided-app-1",
+                  "url": "https://marketplace.firefox.com/app/21066c1c-e45d-4538-8f80-96981df51468/manifest.webapp"
```
Comment (mstriemer) [inline docs/api/topics/latecustomization.rst]: It would probably be useful to include the type of the object as well. `"type": "app"` or something.

File: docs/api/topics/latecustomization.rst [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+  Set the late-customization list for a given region and carrier. Replaces
+  existing list.
+
+  **Request**
+
+  :param app: A webapp slug, if item is a webapp.
+  :type app: str
+  :param extension: An extension slug, if item is an extension.
+  :type app: str
```
Comment (mstriemer) [inline docs/api/topics/latecustomization.rst]: Should this be `:type extension: str`?

File: mkt/feed/views.py [hunk 13]
```diff
@@ -45,6 +45,7 @@
 
 from .fields import DataURLImageField, ImageURLField
 from .models import FeedApp, FeedBrand, FeedCollection, FeedItem, FeedShelf
+
```
Comment (mstriemer) [inline mkt/feed/views.py]: This doesn't look like it should be here.

**Comment by mstriemer** [comment]
The API looks pretty limited right now, we'll likely need to add more fields to be able to correctly display it on the frontend.

r+wc

**Comment by mstriemer** [comment]
I had to add the ID for the item to the response so that I could delete them in the management API. In the management API I just hit the regular app URL to get the data for each app but that isn't ideal for a phone.

File: docs/api/topics/latecustomization.rst [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                  ...
+              }
+              ...
+          ]
+      }
+
+
+.. http:post:: /api/v2/latecustomization/
+
+  Set the late-customization list for a given region and carrier. Replaces
```
Comment (mstriemer) [inline docs/api/topics/latecustomization.rst]: This just creates a single item doesn't it? I don't see how it could replace the existing list. The management portion is currently setup to manage individual items.

Comment (washort) [inline docs/api/topics/latecustomization.rst]: No, you're right, that creates a single item, I didn't update the docs after I changed that. :)

File: mkt/latecustomization/serializers.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
+from rest_framework import relations, serializers
+
+import mkt.carriers
+import mkt.regions
+from mkt.api.fields import SlugChoiceField
+from mkt.extensions.serializers import ExtensionSerializer
+from mkt.fireplace.serializers import FireplaceAppSerializer
```
Comment (mstriemer) [inline mkt/latecustomization/serializers.py]: I thought this serializer would work but it doesn't include `regions` which is required by the transonic app widget. Can you switch this to use `mkt.webapps.serializers.SimpleAppSerializer` please?
